### PR TITLE
Create custom Trim parameter class and overload the trim methods

### LINKED ISSUE #3803: Redis Stream XTRIM with MINID strategy is not fully supported: incompatible parameter type
**Is your feature request related to a problem? Please describe.**
The API to trim a redis stream ([see the correspoding XTRIM command](https://redis.io/commands/xtrim)) supports different strategies via the [org.redisson.api.stream.TrimStrategy](https://javadoc.io/doc/org.redisson/redisson/latest/org/redisson/api/stream/TrimStrategy.html) type.
Yet the additional command parameters in the overloaded methods (e.g. trim, trimAsync, trimNonStrictAsync) will allow only `int`s.
Specifically you can now only pass a
* `int threshold` (to say how many messages you want to keep)
* `int limit` (to remove up to a given number of messages)

Those are inconsistent with the `TrimStrategy.MINID` which expects a message Id as input.

Since the message Ids are composed of a timestamp in millis from epoch and a number, the current `int` param won't be able to technically be used for such a job, even ignoring the semantics of the param.
An `int` is not enough to encode a valid millis timestamp which is part of the message Id.

**Describe the solution you'd like**
It would be nice to provide overloaded methods taking an instance of [org.redisson.api.StreamMessageId](https://github.com/redisson/redisson/blob/master/redisson/src/main/java/org/redisson/api/StreamMessageId.java), and - most importantly- a simple `String`, to allow partial matches.

**Describe alternatives you've considered**
I will probably use a work-around by calling the low-level redis client, but the issue is mostly about the current API which is a bit misleading in what it provides.

### PR DESCRIPTION
**Pull request by ivanopagano**
Fixes #3803 

Deprecate the existing XTRIM methods available in any interface where the TrimStrategy is supported, and overload with one
that accepts a generic TrimParam.

This allows to pass in different types than the current `int` values, to support the full semantics of MINID.

Such TrimParam can only be created via smart constructors, to improve on type safety and user ergonomics.

**CAVEAT**

Currently I can't manage to run the test suite, as explained in the related issue.
I'll report here the problem I have

---

I'm locally testing my fix, but I can't seem to properly run the overall tests as per instructions.

Essentially it looks like my mvn installation can't properly pick up the argLine property to feed to the tests, as the default redis binary path is not properly overridden.

This is the starting log when I try to run the mvn tests as per contribution's instructions
```
Apache Maven 3.8.2 (ea98e05a04480131370aa0c110b8c54cf726c06f)                                                                                                                                                                                                                  
Maven home: /usr/local/Cellar/maven/3.8.2/libexec                                                                                                                                                                                                                              
Java version: 11.0.11, vendor: AdoptOpenJDK, runtime: /Users/ivanopagano/.sdkman/candidates/java/11.0.11.hs-adpt                                                                                                                                                               
Default locale: en_IT, platform encoding: UTF-8                                                                                                                                                                                                                                
OS name: "mac os x", version: "11.5", arch: "x86_64", family: "mac"
```

Are there specific jdk/mvn versions settings required to run the project, apart from what's directly configured via the pom file?

### CODE REVIEW

**Comment by mrniko** [comment]
Thank you for proposed changes.

Did you try this example https://github.com/redisson/redisson/blob/master/.github/CONTRIBUTING.md#running-the-tests

**Comment by ivanopagano** [comment]
> Thank you for proposed changes.
> 
> Did you try this example https://github.com/redisson/redisson/blob/master/.github/CONTRIBUTING.md#running-the-tests

It's exactly what fails to me in the way I showed in the description...

**Comment by ivanopagano** [comment]
> I run it successfully. Windows

On this PR branch? Then I'd deem it's working...

I can't even run the `master` on my macos. First thing that fails is that mvn doesn't pick up the argLine values, hence the test environment can't properly figure out my local executable path.

Additionally, even if I override the hardcoded default to point it to my local redis-server the tests start fine, but I get multiple failures...
Could it be that the main tests are tailored to some specifics of the running env, like OS, mvn, jdk... ?

In the end, if _at least someone_ have run this branch successfully, is it considered good to go?

**Review by mrniko — CHANGES_REQUESTED** [review]

File: redisson/src/main/java/org/redisson/api/RStream.java [hunk 1]
```diff
@@ -768,15 +768,27 @@
...
     long trim(TrimStrategy strategy, int threshold);
 
+    /**
+     * Trims stream with a custom strategy
+     *
+     * @param strategy - trim strategy
+     * @param trimParam - a param specific for the strategy
+     * @return number of deleted messages
+     */
+    long trim(TrimStrategy strategy, TrimParam param);
```
[the remark this answers is not in the document]
Comment (mrniko) [reply redisson/src/main/java/org/redisson/api/RStream.java:787]: I have an idea how to extend API. Below is an example:

```java
trim().maxLen(10).noLimit();
trimStrict().minId(new StreamMessageId(123123));
```

[the remark this answers is not in the document]
Comment (mrniko) [reply redisson/src/main/java/org/redisson/api/RStream.java:787]: Will implement that.

[the remark this answers is not in the document]
Comment (ivanopagano) [reply redisson/src/main/java/org/redisson/api/RStream.java:787]: You wanna build a fluent api to have properly typed inputs?

Sounds a good idea, even tho a bit of rework. I tried to be as non-intrusive as possible. 😁

[the remark this answers is not in the document]
Comment (mrniko) [reply redisson/src/main/java/org/redisson/api/RStream.java:787]: Didn't know how to make fluent api for trim. But now I know. :)

**Comment by mrniko** [comment]
Fixed in https://github.com/redisson/redisson/commit/b9ed3fc743f7d8d14aa0d43e829ac90e6e12ddb3

**Pull request closed without merging**